base/util.jl

```diff
@@ -634,46 +661,57 @@ end
         if isexpr(item, :block)
-            extract_names_and_defvals_from_kwdef_fieldblock!(item, names, defvals)
+            extract_names_types_and_defvals_from_kwdef_fieldblock!(item, names, types, defvals)
         elseif item isa Expr && item.head in (:escape, :var"hygienic-scope")
             n = length(names)
-            extract_names_and_defvals_from_kwdef_fieldblock!(item, names, defvals)
+            extract_names_types_and_defvals_from_kwdef_fieldblock!(item, names, types, defvals)
             for j in n+1:length(defvals)
                 if !isnothing(defvals[j])
                     defvals[j] = Expr(item.head, defvals[j])
                 end
             end
         else
-            def, name, defval = @something(def_name_defval_from_kwdef_fielddef(item), continue)
+            def, name, type, defval = @something(def_name_type_defval_from_kwdef_fielddef(item), continue)
             block.args[i] = def
             push!(names, name)
+            push!(types, type)
             push!(defvals, defval)
         end
     end
 end
 
-function def_name_defval_from_kwdef_fielddef(kwdef)
+function def_name_type_defval_from_kwdef_fielddef(kwdef)
     if kwdef isa Symbol
-        return kwdef, kwdef, nothing
+        return kwdef, kwdef, Any, nothing
     elseif isexpr(kwdef, :(::))
-        name, _ = kwdef.args
-        return kwdef, Meta.unescape(name), nothing
+        name, type = kwdef.args
+        return kwdef, Meta.unescape(name), type, nothing
     elseif isexpr(kwdef, :(=))
         lhs, rhs = kwdef.args
-        def, name, _ = @something(def_name_defval_from_kwdef_fielddef(lhs), return nothing)
-        return def, name, rhs
+        def, name, type, _ = @something(def_name_type_defval_from_kwdef_fielddef(lhs), return nothing)
+        return def, name, type, rhs
     elseif kwdef isa Expr && kwdef.head in (:const, :atomic)
-        def, name, defval = @something(def_name_defval_from_kwdef_fielddef(kwdef.args[1]), return nothing)
-        return Expr(kwdef.head, def), name, defval
+        def, name, type, defval = @something(def_name_type_defval_from_kwdef_fielddef(kwdef.args[1]), return nothing)
+        return Expr(kwdef.head, def), name, type, defval
     elseif kwdef isa Expr && kwdef.head in (:escape, :var"hygienic-scope")
-        def, name, defval = @something(def_name_defval_from_kwdef_fielddef(kwdef.args[1]), return nothing)
-        return Expr(kwdef.head, def), name, isnothing(defval) ? defval : Expr(kwdef.head, defval)
+        def, name, type, defval = @something(def_name_type_defval_from_kwdef_fielddef(kwdef.args[1]), return nothing)
+        return Expr(kwdef.head, def), name, type, isnothing(defval) ? defval : Expr(kwdef.head, defval)
     end
 end
 
+function compute_deftypes(fieldtypes, defvals)
+    Dict(map(unique(fieldtypes)) do sym
+        idxs = findall(==(sym), fieldtypes)
+        isempty(idxs) && return sym => Union{}
+        deftypes = typeof.(eval.(filter(!isnothing, defvals[idxs])))
```

Comment:
I am aware that usage of `eval` is discouraged, so I am happy to learn an alternate way to get the type of the vector if you have something like
```julia
Base.@kwdef struct A{T}
         a::T = [1]
         b::T = [1]
         c::T = [1]
end
```

Reply:
I mean essentially the question that needs to be decided is: What do we choose for the default parameter in the case of `A2`?

The options are:
1. use the type of the first argument to the constructor
```julia
function A(; a = [1.0], b = [1], c = [1])
    A{typeof(a)}(a, b, c)
end
```
2. use the type of the first given default value in the definition
```julia
function A(; a = [1.0], b = [1], c = [1])
    A{Vector{Float64}}(a, b, c)
end
```

For both options we can choose to check whether all types in the definition match and then only do it or not.
Currently I chose the safest route of the second option with the check. But I am happy to ship any other version. Maybe something that could be discussed.